### DISCUSSION TITLE
"Files" relational field

### DISCUSSION THREAD

**Discussion by artecoop** [directus/directus#15647] [Archive]
Hello all!
I've added to an existing data model a "Files" relational fields with the name of "attachments".
When querying the endpoint via rest API, the data returned for the field "attachments" is an array of integers. 
The same result is returned even if I put "attachments.*" on the `fields`param when querying.

How can I have an asset id, or a complex object filled in that field when querying the collection?

Thanks

**Comment by br41nslug** [comment]
Because the files interface makes use of a M2M you may need to use something along the lines of `attachments.directus_files_id.*` in your `fields` assuming the field is called attachments.

**Reply by artecoop** [reply]
That was it, thanks for your help. Docs are very unclear about it!

**Reply by artecoop** [reply]
Hey sorry to bother you. I can't figure out how to upload a file with this configuration using configuration. 
Reading the docs (specifically https://docs.directus.io/reference/files.html), I assumed the flow is to first upload the file, then with the resulting json update the attachment field in my collection's item. But to no avail.

What I'm doing wrong?

**Reply by br41nslug** [reply]
That is the right flow. How are you uploading the file? Have you followed this example https://docs.directus.io/reference/files.html#upload-a-file

**Reply by artecoop** [reply]
Uploading the file is easy, and I can upload freely any file I want.
The problem is associating the uploaded file to a new attachment for the collection. The collection name is `Contract` and contains the `files` field with the structure you kindly explained to me.

After uploading the file I receive the `file object`, exactly as described by the docs. Some code would be similar to this:

```ts
const formData = new FormData();
formData.append('file', payload);

const result = await apiPost<{ data: DirectusFile }>('https://XXXXX.directus.app/files', formData);
contract.attachments.push({ directus_files_id: result.data });

await apiPatch('https://XXXXX.directus.app/items/contract/1250', contract.attachments);
```
where contract is the collections' item I've queried previously.

Patching result in a succesful response, but the item in the collection is not updated, lacking the file relation I've just completed

**Reply by br41nslug** [reply]
Because you're using the Files interface which is an M2M relation i'll stick with the `create/update/delete` structure for readability.
For this example i'll add the newly uploaded file to the `contract` collection:
```js
...
const result = await apiPost<{ data: DirectusFile }>('https://XXXXX.directus.app/files', formData);

contract.attachments = {
  "create": [{ directus_files_id: result.data }],
};

await apiPatch('https://XXXXX.directus.app/items/contract/1250', contract);
```
https://docs.directus.io/reference/introduction.html#one-to-many-many-to-many

**Reply by artecoop** [reply]
Perfect. One last question:

Using the same logic, in order to delete an attachment I need to pass the ID of the file object, which seems to be an UUID. 
But none of these examples works:

```ts
contract.attachments = {
  "delete": [{ directus_files_id: file_id }],
};
```

Error: `Invalid one-to-many update structure: "delete[0]" does not match any of the allowed types`

Same error with

```ts
contract.attachments = {
  "delete": [{ file_id }],
};
```

With 
```ts
contract.attachments = {
  "delete": [file_id],
};
```

returns a success, but without deleting any relation nor file.

If instead I use the numeric ID of the relation, returned by the contract query without specifying the  relation (removing `'attachments.directus_files_id.*'` from `{ fields: ['*', 'attachments.directus_files_id.*'}`) I can successfuly delete both the file and the relation (as expected, I've set it up on the relation triggers in the field).

But specifying the relation in the fields hides the numeric ids. Any help?

**Reply by br41nslug** [reply]
A delete (and update) requires the ID of the junction record as there technically can be more than one of the same file attached to this collection. To get these IDs you will need to fetch the contract with `fields: [...other..., 'attachments.id']`.

**Reply by artecoop** [reply]
Perfect, just perfect. Thank you!!!
